File: morkato/family.py
from __future__ import annotations
from .user import UserTypeFlags
from .utils import NoNullDict
from typing_extensions import Self
from typing import (
  TYPE_CHECKING,
  SupportsInt,
  Optional,
  Dict
)
if TYPE_CHECKING:
  from .types import (
    Family as FamilyPayload,
    NpcType
  )
  from .state import MorkatoConnectionState
  from .ability import Ability
  from .abc import Snowflake
  from .guild import Guild

class Family:
  def __init__(self, state: MorkatoConnectionState, guild: Guild, payload: FamilyPayload) -> None:
    self.state = state
    self.http = state.http
    self.guild = guild
    self.id = int(payload["id"])
    self.from_payload(payload)
  def from_payload(self, payload: FamilyPayload) -> None:
    self.name = payload["name"]
    self.percent = payload["percent"]
    self.user_type = UserTypeFlags(payload["user_type"])
    self.description = payload["description"]
    self.banner = payload["banner"]
  async def update(
    self, *,
    name: Optional[str] = None,
    npc_type: Optional[SupportsInt] = None,
    description: Optional[str] = None,
    banner: Optional[str] = None
  ) -> Self:
    payload = NoNullDict(
      name=name,
      npc_type=npc_type,
      description=description,
      banner=banner
    )
    if payload:
      payload = await self.http.update_family(self.guild.id, self.id, **payload)
      self.from_payload(payload)
    return self
  async def delete(self) -> Self:
    payload = await self.http.delete_family(self.guild.id, self.id)
    self.from_payload(payload)
    self.guild.families.remove(self)
    return self

File: api/src/main/kotlin/morkato/api/model/family/Family.kt
package morkato.api.model.family

import morkato.api.infra.repository.FamilyRepository
import morkato.api.model.guild.Guild
import org.jetbrains.exposed.sql.ResultRow
import java.math.BigDecimal

class Family(
  val guild: Guild,
  val id: Long,
  val name: String,
  val percent: BigDecimal,
  val userType: Int,
  val description: String?,
  val banner: String?
) {
  public constructor(guild: Guild, row: ResultRow) : this(guild, FamilyRepository.FamilyPayload(row));
  public constructor(guild: Guild, payload: FamilyRepository.FamilyPayload) : this(
    guild,
    payload.id,
    payload.name,
    payload.percent,
    payload.userType,
    payload.description,
    payload.banner
  );
  fun update(
    name: String?,
    percent: BigDecimal?,
    userType: Int?,
    description: String?,
    banner: String?
  ) : Family {
    FamilyRepository.updateFamily(
      guildId = this.guild.id,
      id = this.id,
      name = name,
      percent = percent,
      userType = userType,
      description = description,
      banner = banner
    )
    return Family(
      guild = this.guild,
      id = id,
      name = name ?: this.name,
      percent = percent ?: this.percent,
      userType = userType ?: this.userType,
      description = description ?: this.description,
      banner = banner ?: this.banner
    )
  }
  fun delete() : Unit {
    FamilyRepository.deleteFamily(this.guild.id, this.id)
  }
}